### PR TITLE
Update CRMP app

### PR DESCRIPTION
**Pull request by rod-glover**
This PR updates `crmp_app`, which provides the BC Station Data - PCDS portal. 

This PR adds testing for `crmp_app`, based on the infrastructure established for testing `canada_ex_app`. Both the changes to the app and the new test code are quite simple.

### CODE REVIEW

**Comment by corviday** [comment]
A bunch of the crmp tests are failing for me, with the error message ` ReferenceError: XMLSerializer is not defined` - a modulel that should have been imported into the test suite but wasn't?

File: pdp/static/js/crmp_app.js [hunk 1]
```diff
@@ -2,35 +2,41 @@
 /*global $, jQuery, pdp, map, init_crmp_map, CRMPFilterChange, getCRMPControls, getCRMPDownloadOptions, getActiveFilters, polyChange, downloadMetadata*/
 "use strict";
 
-var map;
+// var map;
```
Comment (corviday) [inline pdp/static/js/crmp_app.js]: If this is an explanatory comment, I do not understand what it means.

File: pdp/static/js/crmp_controls.js [hunk 2]
```diff
@@ -136,8 +136,22 @@ function getResetButton(map) {
...
+    var calendar = calendars['gregorian'];
+    var units = 'days';
+    var cfTimeSystem = new calendars.CfTimeSystem(
+        units,
+        new calendars.CalendarDatetime(calendar, 1870, 1, 1),
+        Math.floor((2100 - 1870 + 1) * 365.2425)
+    );
+    var startDate = cfTimeSystem.firstCfDatetime();
+    var endDate = cfTimeSystem.todayAsCfDatetime();
+
```
Comment (corviday) [inline pdp/static/js/crmp_controls.js]: When I run this portal on my work station, the actual date range seems to be 1870-01-01 to 2100-01-01, not 1870-01-01 to today. It lets me set the end date to, say, 2050-06-21 without complain.

daterange

**Comment by rod-glover** [comment]
@corviday , this works on my workstation now. Fixed a couple of stupids. Also fixed the problems in tests.

**Comment by rod-glover** [comment]
[Demo on docker](http://docker-dev02.pcic.uvic.ca:30509/pcds/map/).

Station popups don't work due to CORS issues. I'm pretty sure that is just due to running on docker-dev02 and not due to a code change.

**Comment by corviday** [comment]
I don't know what to make of this. I get the following errors when running the tests:

```
 PASS  pdp/static/js/__test__/condExport.test.js
  ✓ it works (3ms)

  console.error node_modules/jest-jasmine2/build/jasmine/Env.js:289
    Unhandled error

  console.error node_modules/jest-jasmine2/build/jasmine/Env.js:290
    TypeError: Cannot read property '_ownerDocument' of undefined
        at innerInvokeEventListeners (/home/lzeman/tmp/pdp/node_modules/jest-environment-jsdom-fourteen/node_modules/jsdom/lib/jsdom/living/events/EventTarget-impl.js:325:26)
        at invokeEventListeners (/home/lzeman/tmp/pdp/node_modules/jest-environment-jsdom-fourteen/node_modules/jsdom/lib/jsdom/living/events/EventTarget-impl.js:266:3)
        at EventTargetImpl._dispatch (/home/lzeman/tmp/pdp/node_modules/jest-environment-jsdom-fourteen/node_modules/jsdom/lib/jsdom/living/events/EventTarget-impl.js:212:11)
        at fireAnEvent (/home/lzeman/tmp/pdp/node_modules/jest-environment-jsdom-fourteen/node_modules/jsdom/lib/jsdom/living/helpers/events.js:17:36)
        at Document.window.document.addEventListener (/home/lzeman/tmp/pdp/node_modules/jest-environment-jsdom-fourteen/node_modules/jsdom/lib/jsdom/browser/Window.js:639:9)
        at innerInvokeEventListeners (/home/lzeman/tmp/pdp/node_modules/jest-environment-jsdom-fourteen/node_modules/jsdom/lib/jsdom/living/events/EventTarget-impl.js:315:27)
        at invokeEventListeners (/home/lzeman/tmp/pdp/node_modules/jest-environment-jsdom-fourteen/node_modules/jsdom/lib/jsdom/living/events/EventTarget-impl.js:266:3)
        at DocumentImpl._dispatch (/home/lzeman/tmp/pdp/node_modules/jest-environment-jsdom-fourteen/node_modules/jsdom/lib/jsdom/living/events/EventTarget-impl.js:212:11)
        at fireAnEvent (/home/lzeman/tmp/pdp/node_modules/jest-environment-jsdom-fourteen/node_modules/jsdom/lib/jsdom/living/helpers/events.js:17:36)
        at dispatchEvent (/home/lzeman/tmp/pdp/node_modules/jest-environment-jsdom-fourteen/node_modules/jsdom/lib/jsdom/living/nodes/Document-impl.js:508:9)

  console.error node_modules/jest-environment-jsdom-fourteen/node_modules/jsdom/lib/jsdom/virtual-console.js:29
    Error: Uncaught [TypeError: Cannot read property '_ownerDocument' of undefined]
        at reportException (/home/lzeman/tmp/pdp/node_modules/jest-environment-jsdom-fourteen/node_modules/jsdom/lib/jsdom/living/helpers/runtime-script-errors.js:62:24)
        at innerInvokeEventListeners (/home/lzeman/tmp/pdp/node_modules/jest-environment-jsdom-fourteen/node_modules/jsdom/lib/jsdom/living/events/EventTarget-impl.js:331:9)
        at invokeEventListeners (/home/lzeman/tmp/pdp/node_modules/jest-environment-jsdom-fourteen/node_modules/jsdom/lib/jsdom/living/events/EventTarget-impl.js:266:3)
        at DocumentImpl._dispatch (/home/lzeman/tmp/pdp/node_modules/jest-environment-jsdom-fourteen/node_modules/jsdom/lib/jsdom/living/events/EventTarget-impl.js:212:11)
        at fireAnEvent (/home/lzeman/tmp/pdp/node_modules/jest-environment-jsdom-fourteen/node_modules/jsdom/lib/jsdom/living/helpers/events.js:17:36)
        at dispatchEvent (/home/lzeman/tmp/pdp/node_modules/jest-environment-jsdom-fourteen/node_modules/jsdom/lib/jsdom/living/nodes/Document-impl.js:508:9)
        at resolve (/home/lzeman/tmp/pdp/node_modules/jest-environment-jsdom-fourteen/node_modules/jsdom/lib/jsdom/living/nodes/Document-impl.js:513:11)
        at new Promise (<anonymous>)
        at onLoad (/home/lzeman/tmp/pdp/node_modules/jest-environment-jsdom-fourteen/node_modules/jsdom/lib/jsdom/living/nodes/Document-impl.js:511:14)
        at Object.check (/home/lzeman/tmp/pdp/node_modules/jest-environment-jsdom-fourteen/node_modules/jsdom/lib/jsdom/browser/resources/resource-queue.js:76:23) TypeError: Cannot read property '_ownerDocument' of undefined
        at innerInvokeEventListeners (/home/lzeman/tmp/pdp/node_modules/jest-environment-jsdom-fourteen/node_modules/jsdom/lib/jsdom/living/events/EventTarget-impl.js:325:26)
        at invokeEventListeners (/home/lzeman/tmp/pdp/node_modules/jest-environment-jsdom-fourteen/node_modules/jsdom/lib/jsdom/living/events/EventTarget-impl.js:266:3)
        at EventTargetImpl._dispatch (/home/lzeman/tmp/pdp/node_modules/jest-environment-jsdom-fourteen/node_modules/jsdom/lib/jsdom/living/events/EventTarget-impl.js:212:11)
        at fireAnEvent (/home/lzeman/tmp/pdp/node_modules/jest-environment-jsdom-fourteen/node_modules/jsdom/lib/jsdom/living/helpers/events.js:17:36)
        at Document.window.document.addEventListener (/home/lzeman/tmp/pdp/node_modules/jest-environment-jsdom-fourteen/node_modules/jsdom/lib/jsdom/browser/Window.js:639:9)
        at innerInvokeEventListeners (/home/lzeman/tmp/pdp/node_modules/jest-environment-jsdom-fourteen/node_modules/jsdom/lib/jsdom/living/events/EventTarget-impl.js:315:27)
        at invokeEventListeners (/home/lzeman/tmp/pdp/node_modules/jest-environment-jsdom-fourteen/node_modules/jsdom/lib/jsdom/living/events/EventTarget-impl.js:266:3)
        at DocumentImpl._dispatch (/home/lzeman/tmp/pdp/node_modules/jest-environment-jsdom-fourteen/node_modules/jsdom/lib/jsdom/living/events/EventTarget-impl.js:212:11)
        at fireAnEvent (/home/lzeman/tmp/pdp/node_modules/jest-environment-jsdom-fourteen/node_modules/jsdom/lib/jsdom/living/helpers/events.js:17:36)
        at dispatchEvent (/home/lzeman/tmp/pdp/node_modules/jest-environment-jsdom-fourteen/node_modules/jsdom/lib/jsdom/living/nodes/Document-impl.js:508:9)

  console.error node_modules/jest-jasmine2/build/jasmine/Env.js:289
    Unhandled error

  console.error node_modules/jest-jasmine2/build/jasmine/Env.js:290
    TypeError: Cannot read property '_ownerDocument' of undefined
        at innerInvokeEventListeners (/home/lzeman/tmp/pdp/node_modules/jest-environment-jsdom-fourteen/node_modules/jsdom/lib/jsdom/living/events/EventTarget-impl.js:325:26)
        at invokeEventListeners (/home/lzeman/tmp/pdp/node_modules/jest-environment-jsdom-fourteen/node_modules/jsdom/lib/jsdom/living/events/EventTarget-impl.js:266:3)
        at EventTargetImpl._dispatch (/home/lzeman/tmp/pdp/node_modules/jest-environment-jsdom-fourteen/node_modules/jsdom/lib/jsdom/living/events/EventTarget-impl.js:212:11)
        at fireAnEvent (/home/lzeman/tmp/pdp/node_modules/jest-environment-jsdom-fourteen/node_modules/jsdom/lib/jsdom/living/helpers/events.js:17:36)
        at Document.window.document.addEventListener (/home/lzeman/tmp/pdp/node_modules/jest-environment-jsdom-fourteen/node_modules/jsdom/lib/jsdom/browser/Window.js:639:9)
        at innerInvokeEventListeners (/home/lzeman/tmp/pdp/node_modules/jest-environment-jsdom-fourteen/node_modules/jsdom/lib/jsdom/living/events/EventTarget-impl.js:315:27)
        at invokeEventListeners (/home/lzeman/tmp/pdp/node_modules/jest-environment-jsdom-fourteen/node_modules/jsdom/lib/jsdom/living/events/EventTarget-impl.js:266:3)
        at DocumentImpl._dispatch (/home/lzeman/tmp/pdp/node_modules/jest-environment-jsdom-fourteen/node_modules/jsdom/lib/jsdom/living/events/EventTarget-impl.js:212:11)
        at fireAnEvent (/home/lzeman/tmp/pdp/node_modules/jest-environment-jsdom-fourteen/node_modules/jsdom/lib/jsdom/living/helpers/events.js:17:36)
        at dispatchEvent (/home/lzeman/tmp/pdp/node_modules/jest-environment-jsdom-fourteen/node_modules/jsdom/lib/jsdom/living/nodes/Document-impl.js:508:9)

  console.error node_modules/jest-environment-jsdom-fourteen/node_modules/jsdom/lib/jsdom/virtual-console.js:29
    Error: Uncaught [TypeError: Cannot read property '_ownerDocument' of undefined]
        at reportException (/home/lzeman/tmp/pdp/node_modules/jest-environment-jsdom-fourteen/node_modules/jsdom/lib/jsdom/living/helpers/runtime-script-errors.js:62:24)
        at innerInvokeEventListeners (/home/lzeman/tmp/pdp/node_modules/jest-environment-jsdom-fourteen/node_modules/jsdom/lib/jsdom/living/events/EventTarget-impl.js:331:9)
        at invokeEventListeners (/home/lzeman/tmp/pdp/node_modules/jest-environment-jsdom-fourteen/node_modules/jsdom/lib/jsdom/living/events/EventTarget-impl.js:266:3)
        at DocumentImpl._dispatch (/home/lzeman/tmp/pdp/node_modules/jest-environment-jsdom-fourteen/node_modules/jsdom/lib/jsdom/living/events/EventTarget-impl.js:212:11)
        at fireAnEvent (/home/lzeman/tmp/pdp/node_modules/jest-environment-jsdom-fourteen/node_modules/jsdom/lib/jsdom/living/helpers/events.js:17:36)
        at dispatchEvent (/home/lzeman/tmp/pdp/node_modules/jest-environment-jsdom-fourteen/node_modules/jsdom/lib/jsdom/living/nodes/Document-impl.js:508:9)
        at resolve (/home/lzeman/tmp/pdp/node_modules/jest-environment-jsdom-fourteen/node_modules/jsdom/lib/jsdom/living/nodes/Document-impl.js:513:11)
        at new Promise (<anonymous>)
        at onLoad (/home/lzeman/tmp/pdp/node_modules/jest-environment-jsdom-fourteen/node_modules/jsdom/lib/jsdom/living/nodes/Document-impl.js:511:14)
        at Object.check (/home/lzeman/tmp/pdp/node_modules/jest-environment-jsdom-fourteen/node_modules/jsdom/lib/jsdom/browser/resources/resource-queue.js:76:23) TypeError: Cannot read property '_ownerDocument' of undefined
        at innerInvokeEventListeners (/home/lzeman/tmp/pdp/node_modules/jest-environment-jsdom-fourteen/node_modules/jsdom/lib/jsdom/living/events/EventTarget-impl.js:325:26)
        at invokeEventListeners (/home/lzeman/tmp/pdp/node_modules/jest-environment-jsdom-fourteen/node_modules/jsdom/lib/jsdom/living/events/EventTarget-impl.js:266:3)
        at EventTargetImpl._dispatch (/home/lzeman/tmp/pdp/node_modules/jest-environment-jsdom-fourteen/node_modules/jsdom/lib/jsdom/living/events/EventTarget-impl.js:212:11)
        at fireAnEvent (/home/lzeman/tmp/pdp/node_modules/jest-environment-jsdom-fourteen/node_modules/jsdom/lib/jsdom/living/helpers/events.js:17:36)
        at Document.window.document.addEventListener (/home/lzeman/tmp/pdp/node_modules/jest-environment-jsdom-fourteen/node_modules/jsdom/lib/jsdom/browser/Window.js:639:9)
        at innerInvokeEventListeners (/home/lzeman/tmp/pdp/node_modules/jest-environment-jsdom-fourteen/node_modules/jsdom/lib/jsdom/living/events/EventTarget-impl.js:315:27)
        at invokeEventListeners (/home/lzeman/tmp/pdp/node_modules/jest-environment-jsdom-fourteen/node_modules/jsdom/lib/jsdom/living/events/EventTarget-impl.js:266:3)
        at DocumentImpl._dispatch (/home/lzeman/tmp/pdp/node_modules/jest-environment-jsdom-fourteen/node_modules/jsdom/lib/jsdom/living/events/EventTarget-impl.js:212:11)
        at fireAnEvent (/home/lzeman/tmp/pdp/node_modules/jest-environment-jsdom-fourteen/node_modules/jsdom/lib/jsdom/living/helpers/events.js:17:36)
        at dispatchEvent (/home/lzeman/tmp/pdp/node_modules/jest-environment-jsdom-fourteen/node_modules/jsdom/lib/jsdom/living/nodes/Document-impl.js:508:9)
```

The console is reporting an error, but the test passes. Not sure if anything needs to be fixed here or not.

**Comment by corviday** [comment]
Putting in an invalid date causes the map to be replaced by a bunch of broken images. 

That isn't a direct result of this PR, though, and it's probably outside the scope of the PR. This particular portal does not handle errors correctly and will fire off malformed requests to ncWMS if any of the map attributes are affected by an error; [the map is also replaced by a bunch of broken image symbols if you enter a polygon consisting of only two points](https://github.com/pacificclimate/pdp/issues/92#issuecomment-417486434).

We need an "improve error handling on the CRMP portal" issue, but since it will be obvious to users what's gone wrong in this case (since there's also an error message under the date in the case of invalid dates) it doesn't seem urgent to fix the issue in this PR.

**Comment by corviday** [comment]
The previous PR, for the Downscaled GCMS Portal, would allow you to enter an not fully specified date, like just putting in 1970 and 2000. This one does not. Any particular reason for the different behavior?

**Comment by corviday** [comment]
A bit of UI oddness: this doesn't seem to be true with the current live portal, but pressing enter after typing a date in the date entry boxes initiates a download. 

I was surprised by that behavior - I expected pressing enter to confirm the date entry, and then to click either "Climatology" or "Timeseries" to actually download data. But maybe I'm not a typical user and this is what a typical user would expect.

**Comment by rod-glover** [comment]
@corviday , thanks for your as-always careful review. Some replies:

- Errors running tests: Yeah, I know about that. I've tried to fix it but it's tenacious, and I believe it is one of the little surprises that come with Jest (which in general is an excellent package, but some of its plugins produce unwanted if harmless side effects like this one). I'll have another go at fixing it.

- Invalid date breaks map: I agree, not a problem particular to this PR.

- Partial date breaks map: A partial date is formally valid (and it doesn't trigger a date validation error). But it does break the map, which is wrong. It should be one or the other! I will see about making it work with partial dates.

- Return behaviour: That is standard behaviour for HTML forms. It is clearly not desired behaviour here. I have added an [issue](https://github.com/pacificclimate/pdp/issues/126) for this.

**Comment by rod-glover** [comment]
@corviday : 

- Partial date problem fixed: Partial dates are expanded to full dates, so that the thingy that gets the date out of the input field gets a date string it likes, and so that the user can see what the program is doing based on their partial date. LMK what you think of this solution.

- No luck on the ugly errors thrown by tests. I don't think it's worth the time to fix at the moment.

**Comment by rod-glover** [comment]
Hold on for a moment while I put up a new demo. The previously working docker build now fails. Argh.

**Comment by rod-glover** [comment]
[Demo](http://docker-dev01.pcic.uvic.ca:30509/pcds/map/) finally up. Note this is on docker-dev01 now. Previous link invalid.

**Comment by rod-glover** [comment]
@corviday ^^ if you have a moment. Interested in your feedback on the change to the date input.

**Comment by corviday** [comment]
It works, hooray!

Is it possible to use a similar date-preprocessing technique in cases where someone has entered an invalid date? For example, if I enter 2019/13/01 as the end date and then hit enter, it takes me to [a 500 error](https://services.pacificclimate.org/data/pcds/agg/?from-date=1870%2F01%2F01&to-date=2019%2F13%2F01&input-polygon=&input-var=&network-name=&input-freq=&data-format=nc&download-climatology=Climatology) or a [different 500 error](https://services.pacificclimate.org/data/pcds/agg/?from-date=1870%2F01%2F01&to-date=2019%2F06%2F40&input-polygon=MULTIPOLYGON%28%28%28-134.5240691989438+59.130842681587865%2C-130.90719530058658+59.12784428119393%2C-132.18283321952785+57.49711410875174%2C-134.5240691989438+59.130842681587865%29%29%29&input-var=&network-name=&input-freq=&data-format=nc&download-climatology=Climatology).

I suspect fixing that will have to wait until #126 is fixed, but I wanted to check, just because it's pretty disruptive to the user experience - if you go to the 500 page, you will have lost your polygon and time selections when you hit back.

**Comment by rod-glover** [comment]
What I think should happen is that the download button(s) should be disabled when the dates are not valid.  I think that is best done as a separate [issue](https://github.com/pacificclimate/pdp/issues/127), but included in this epic (series of PRs). That's in addition to #126.

**Pull request merged**